ompi/mca/op/aarch64/op_aarch64_functions.c

```diff
@@ -0,0 +1,641 @@
+ int16_t:   __extension__({ switch (how_much) { DUMP8(out, in1, in2) }}), \
+ uint16_t:  __extension__({ switch (how_much) { DUMP8(out, in1, in2) }}), \
+ int32_t:   __extension__({ switch (how_much) { DUMP4(out, in1, in2) }}), \
+ uint32_t:  __extension__({ switch (how_much) { DUMP4(out, in1, in2) }}), \
+ int64_t:   __extension__({ switch (how_much) { DUMP2(out, in1, in2) }}), \
+ uint64_t:  __extension__({ switch (how_much) { DUMP2(out, in1, in2) }}), \
+ float32_t: __extension__({ switch (how_much) { DUMP4(out, in1, in2) }}), \
+ float64_t: __extension__({ switch (how_much) { DUMP2(out, in1, in2) }}))
+#endif /* defined(GENERATE_SVE_CODE) */
+
+/*
+ * Since all the functions in this file are essentially identical, we
+ * use a macro to substitute in names and types.  The core operation
+ * in all functions that use this macro is the same.
+ *
+ * This macro is for (out op in).
+ *
+ */
+#if defined(GENERATE_NEON_CODE)
+#define OP_AARCH64_FUNC(name, type_name, type_size, type_cnt, type, op)                       \
+    static void OP_CONCAT(ompi_op_aarch64_2buff_##name##_##type##type_size##_t,               \
+                          APPEND)(const void *_in, void *_out, int *count,                    \
+                                  struct ompi_datatype_t **dtype,                             \
+                                  struct ompi_op_base_module_1_0_0_t *module)                 \
+    {                                                                                         \
+        int left_over = *count;                                                               \
+        type##type_size##_t *in = (type##type_size##_t *) _in,                                \
+                            *out = (type##type_size##_t *) _out;                              \
+        OP_CONCAT(OMPI_OP_TYPE_PREPEND, type##type_size####x##type_cnt##_t) vsrc, vdst;       \
+        for (; left_over >= type_cnt; left_over -= type_cnt) {                                \
+            vsrc = vld1q##_##type_name##type_size(in);                                        \
+            vdst = vld1q##_##type_name##type_size(out);                                       \
+            in += type_cnt;                                                                   \
+            vdst = OP_CONCAT(OMPI_OP_OP_PREPEND, op##q##_##type_name##type_size)(vdst, vsrc); \
+            vst1q##_##type_name##type_size(out, vdst);                                        \
+            out += type_cnt;                                                                  \
+        }                                                                                     \
+                                                                                              \
+        while (left_over > 0) {                                                               \
+            int how_much = (left_over > 8) ? 8 : left_over;                                   \
```

Comment:
For NEON, maximum leftover should be 15 (for 1-byte operands).  If we implement that, we wouldn't need the loop at all.

Reply:
If I remove the __extension__ to make the above macro portable, all cases will have a 16 case long switch (except if the compiler is really really smart about it). I looked at the gcc generated assembly and it didn't look very smart to me, so I decided to be more cautious on the macro use.

Reply:
I was playing around with it here: https://godbolt.org/z/1GzMMsjPz

Seems to me it does exactly as we expect: jumps to a label and then falls through as desired to finish the last steps with only 1 branch.  Seems like having a few extra (dead) statements won't hurt us too much, but actually it seems that gcc -O3 has optimized those away for all the versions I checked.  Clang too.

Maybe you were seeing something I missed?  Code seems much cleaner this way.